File: programmers/고득점 Kit/DFSBFS/43164.py
# Solved
def solution(tickets):
    answer = []
    tickets.sort()
    
    def dfs(path):
        if len(path)==len(tickets)+1:
            answer.append(path)
            return
        
        for i, (s, e) in enumerate(tickets):
            if not visited[i] and path[-1]==s:
                visited[i] = True
                dfs(path+[e])
                visited[i] = False
    
    visited = [False] * len(tickets)
    dfs(["ICN"])
    return answer[0] # 사전순이므로 첫 번째 리턴
'''
visited = set() 으로 목적지 e만 방문 체크 
→ 같은 목적지로 가는 여러 티켓 구분 안 됨 → 티켓 중복 사용 방지 실패
'''

# 다른 사람 풀이 -  Hierholzer's Algorithm
def solution(tickets):
    routes = {}
    for t in tickets:
        routes[t[0]] = routes.get(t[0], []) + [t[1]]
    for r in routes:
        routes[r].sort(reverse=True)

    stack = ["ICN"]
    path = []
    while stack:
        top = stack[-1]
        if top in routes and routes[top]:
            stack.append(routes[top].pop())
        else:
            path.append(stack.pop())

    return path[::-1]
'''
dictionary get(찾고자 하는 key값, key가 없을 때 리턴할 default값)

Hierholzer 알고리즘은 오일러 회로(Euler circuit)를 찾는 알고리즘
https://github.com/2018007956/Teamnote-archive/blob/main/Searching/Hierholzer.py
'''

File: programmers/고득점 Kit/DFSBFS/43164.java
import java.util.*;
class Solution {
    List<String> answer = new ArrayList<>();
    boolean[] visited;
    String[][] tickets;

    public String[] solution(String[][] tickets) {
        this.tickets = tickets;
        this.visited = new boolean[tickets.length];

        Arrays.sort(tickets, (a, b) -> {
            if (!a[0].equals(b[0])) return a[0].compareTo(b[0]);
            return a[1].compareTo(b[1]);
        });

        dfs(new ArrayList<>(List.of("ICN")));
        return answer.toArray(new String[0]);
    }

    public void dfs(List<String> path) {
        if(path.size()==tickets.length+1) {
            if(answer.isEmpty()) {
                answer = new ArrayList<>(path);
            }
            return;
        }

        for(int i=0; i<tickets.length; i++) {
            if(!visited[i] && path.get(path.size()-1).equals(tickets[i][0])) {
                visited[i] = true;
                path.add(tickets[i][1]);
                dfs(path);
                path.remove(path.size()-1);
                visited[i] = false;
            }
        }
    }
}

// 다른 사람 풀이
import java.util.ArrayList;
import java.util.Collections;
class Solution {
    static ArrayList<String> answer = new ArrayList<>();
    static boolean[] visited;

    public String[] solution(String[][] tickets) {
        visited = new boolean[tickets.length];
        DFS(0, "ICN", "ICN", tickets);
        Collections.sort(answer);
        String[] temp = answer.get(0).split(" ");
        return temp;
    }

    private static void DFS(int cnt, String icn, String word, String[][] tickets) {
        if (cnt == tickets.length) {
            answer.add(word);
        } else {
            for (int i=0; i<tickets.length; i++) {
                if (!visited[i] && tickets[i][0].equals(icn)) {
                    visited[i] = true;
                    DFS(cnt+1, tickets[i][1], word+" "+tickets[i][1], tickets);
                    visited[i] = false;
                }
            }
        }
    }
}